pkg/p2p/internal/send_chan.go

```diff
@@ -0,0 +1,170 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	proto "github.com/pingcap/tiflow/proto/p2p"
+	"go.uber.org/zap"
+)
+
+// SendChan is a specialized channel used to implement
+// the asynchronous interface of the MessageClient.
+// SendChan is a MPSC channel.
+type SendChan struct {
+	mu      sync.Mutex
+	buf     []*proto.MessageEntry
+	sendIdx int64
+	recvIdx int64
+
+	// hasNewMsg is a buffered channel that the sending goroutine(s)
+	// use to notify the receiver end.
+	hasNewMsg chan struct{}
+
+	// notifyChan is an unbuffered channel that
+	// SendSync writes to in order to determine whether the
+	// consumer is ready.
+	notifyChan chan struct{}
+
+	cap int64
+}
+
+// NewSendChan returns a new SendChan.
+func NewSendChan(cap int64) *SendChan {
+	return &SendChan{
+		buf:        make([]*proto.MessageEntry, cap),
+		hasNewMsg:  make(chan struct{}, 1),
+		notifyChan: make(chan struct{}),
+		cap:        cap,
+	}
+}
+
+// SendSync sends a message synchronously.
+// The performance of SendSync is about 500x slower than
+// SendAsync when CPU is saturated, so please refrain from
+// using it unless necessary or for testing.
+func (c *SendChan) SendSync(
+	ctx context.Context,
+	topic string,
+	value []byte,
+	closeCh <-chan struct{},
+	nextSeq func() int64,
+) (int64, error) {
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if ok, seq := c.SendAsync(topic, value, nextSeq); ok {
+			return seq, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return 0, errors.Trace(ctx.Err())
+		case <-closeCh:
+			return 0, cerror.ErrPeerMessageClientClosed.GenWithStackByArgs()
+		case <-c.notifyChan:
+		case <-ticker.C:
+		}
+	}
+}
+
+// SendAsync tries to send a message. If the message is accepted, nextSeq will be called
+// once, and the returned value will be used as the Sequence number of the message.
+func (c *SendChan) SendAsync(topic string, value []byte, nextSeq func() int64) (ok bool, seq int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.sendIdx-c.recvIdx > c.cap {
+		log.Panic("unreachable",
+			zap.Int64("sendIdx", c.sendIdx),
+			zap.Int64("recvIndex", c.recvIdx))
+	}
+
+	if c.sendIdx-c.recvIdx == c.cap {
+		return false, 0
+	}
+
+	seq = nextSeq()
+	c.buf[c.sendIdx%c.cap] = &proto.MessageEntry{
+		Topic:    topic,
+		Content:  value,
+		Sequence: seq,
+	}
+	c.sendIdx++
+
+	select {
+	case c.hasNewMsg <- struct{}{}:
+	default:
+	}
+
+	return true, seq
+}
+
+// Receive receives one message from the channel.
+// If a message is received from `tick`, the function will return
+// (nil, false, nil).
+func (c *SendChan) Receive(ctx context.Context, tick <-chan time.Time) (*proto.MessageEntry, bool, error) {
```

Comment:
maybe this tick can be replaced by a configurable tick interval, same with `SendSync`?

Reply:
This `tick` is passed from the caller, and is controller by the caller's configuration.